corehq/messaging/util.py

```diff
@@ -41,16 +49,31 @@ def rule_initiation_key_is_set(self):
     def rule_initiation_key_minutes_remaining(self):
         return (self.client.ttl(self.rule_initiation_key) // 60) or 1
 
-    def set_initial_progress(self):
+    def set_initial_progress(self, shard_count=0):
+        # shard_count is passed when tasks are run per each shard
         self.client.set(self.current_key, 0)
         self.client.set(self.total_key, 0)
+        if shard_count:
+            self.client.set(self.shard_count_key, shard_count)
         self.client.expire(self.current_key, self.key_expiry)
         self.client.expire(self.total_key, self.key_expiry)
         self.client.delete(self.rule_cancellation_key)
         self.set_rule_initiation_key()
 
+    def mark_shard_complete(self, db_alias):
+        completed_shards = self.client.get(self.completed_shards_key) or []
+        self.client.set(self.completed_shards_key, completed_shards + [db_alias])
+        self.client.expire(self.completed_shards_key, self.key_expiry)
+
+    def all_shards_completed(self):
+        completed_count = len(self.client.get(self.completed_shards_key) or [])
+        total_count = self.client.get(self.shard_count_key) or 0
+        return completed_count == total_count
+
     def set_rule_complete(self):
         self.clear_rule_initiation_key()
+        for key in [self.shard_count_key, self.completed_shards_key]:
+            self.client.delete(key)
```

Comment:
It's possible we originally thought `run_messaging_rule` would not take more than two hours to complete (the `sync_case_for_messaging_rule` sub-tasks may of course run for many more hours). Maybe the mechanics of `rule_initiation_key` are based on outdated assumptions?

Reply:
This was outdated diff, I am not clearing these. But agreed generally some of the logic assumes this was going to be quick. I am not attempting to fix the bugs around these, since they are mostly about visibility into the task and support some progress related features in the UI (i.e. doesn't play big part in the mechanics of the task itself, except the ones I added)

Reply:
there are definitely more things that need attention in that helper @millerdev 
the method that shares percentage of progress is totally broken, it shows either 0% when it stars and then constantly shows 100%.
But ya this PR is focussed mainly on the processing improvements.